modules/packages/DistributedBag.chpl

```diff
@@ -64,631 +62,791 @@ use BlockDist;
-    coforall segmentIdx in 0 .. #here.maxTaskPar {
-      var nSegmentElems : [localThis.targetLocales.size] int;
-      var locIdx = 0;
-      for loc in localThis.targetLocales do on loc {
-        var nElems = getPrivatizedThis.bag.segments[segmentIdx].nElems.read() : int;
-        nSegmentElems[locIdx] = nElems;
-        locIdx += 1;
+    /*
+      Performs a lookup to determine if the requested element exists in this bag.
+      This method is best-effort and can be non-deterministic for concurrent
+      updates across nodes, and may miss elements resulting from any concurrent
+      insertion or removal operations.
+    */
+    proc contains(elt : eltType) : bool {
+      var foundElt : atomic bool;
+      forall elem in getPrivatizedThis {
+        if elem == elt {
+          foundElt.write(true);
+        }
       }
+      return foundElt.read();
+    }
+
+    /*
+      Clear all bags across all nodes in a best-effort approach. Elements added or
+      moved around from concurrent additions or removals may be missed while clearing.
+    */
+    proc clear() {
+      var localThis = getPrivatizedThis;
+      coforall loc in localThis.targetLocales do on loc {
+        var instance = getPrivatizedThis;
+        forall segmentIdx in 0 .. #here.maxTaskPar {
+          ref segment = instance.bag.segments[segmentIdx];
+          if segment.acquireIfNonEmpty(STATUS_REMOVE) {
+            var block = segment.headBlock;
+            while block != nil {
+              var tmp = block;
+              block = block.next;
+              delete tmp;
+            }
 
-      // Find the average and the excess. The excess is calculated as the amount
-      // of elements a segment has over the average, which is used to calculate
-      // the buffer size for each segment.
-      var total = (+ reduce nSegmentElems);
-      var avg = total / locIdx;
-      var excess : int;
-      for nElems in nSegmentElems {
-        if nElems > avg {
-          excess += nElems - avg;
+            segment.headBlock = nil;
+            segment.tailBlock = nil;
+            segment.nElems.write(0);
+            segment.releaseStatus();
+          }
         }
       }
+    }
 
-      // Allocate buffer, which holds the 'excess' elements for redistribution.
-      // Then fill it.
-      var buffer = c_malloc(eltType, excess);
-      var bufferOffset = 0;
+    /*
+      Triggers a more static approach to load balancing, fairly redistributing all
+      elements fairly for bags across nodes. The result will result in all segments
+      having roughly the same amount of elements.
+      
+      .. note::
+
+        This method is very heavy-weight in that it should not be called too
+        often. Dynamic work stealing handles cases where there is a relatively fair
+        distribution across majority of nodes, but this should be called when you have
+        a severe imbalance, or when you have a smaller number of elements to balance.

```

Comment:
"or when you have a smaller number of elements to balance" sounds a bit contradictory to what you are saying before.